osg-project-waittime/calculate-waittime.py

```diff
@@ -0,0 +1,235 @@
+    userAttrDict = {}
+    for userAttr in users:
+        if userAttr.starttime < parser.parse("2021-03-09"):
+            #print("Before QueueTime available")
+            continue
+        found1000 = False
+        s = generateRawQuery(userAttr.username, userAttr.starttime, userAttr.endtime)
+        s = s.params(preserve_order=True)
+        for record in s.scan():
+            userAttr.njobs += 1
+            userAttr.walltime += record['WallDuration']
+            userAttr.corehours += record['CoreHours']
+            if 'QueueTime' in record:
+                userAttr.queuetime += (parser.parse(record['EndTime']).timestamp() - parser.parse(record['QueueTime']).timestamp()) - record['WallDuration']
+            else:
+                print("QueueTime not found when it should be for probe:{}")
+            # check if we have 1000 hours
+            if userAttr.corehours > 1000:
+                found1000 = True
+                break
+        if found1000:
+            print("Found the user: {} with QueueTime (hours): {} for CoreHours: {}".format(userAttr.username, userAttr.queuetime/HOUR, userAttr.corehours))
+        else:
+            print("Did not find 1000 hours in {} jobs of usage for user: {} in set of days: {}-{}".format(userAttr.njobs, userAttr.username, userAttr.starttime, userAttr.endtime))
+
+    return users
+
+
+
+def add_args():
+
+    argsparser = argparse.ArgumentParser(description='Calculate waittime for users')
+    argsparser.add_argument("starttime", type=str, help="Start Time, for example 2021-03-01")
+    argsparser.add_argument("endtime", type=str, help="End Time, for example 2021-03-31")
+    return argsparser
+
+def main():
+    argsparser = add_args()
+    args = argsparser.parse_args()
+
+    startDatetime = parser.parse(args.starttime)
+    endDatetime = parser.parse(args.endtime)
+
+    # Find all user with any usage in the last 1 month
+    perDay = getUsersPerDay(startDatetime, endDatetime)
+
+    # Gather per-day statistics for each user
+    users = getIdleUsers(perDay)
+
+    # Queue times
+    queueTimes = getQueueTimes(users)
+    columnNames = ["Username", "ProjectName", "Start Time", "End Time", "Days of Zero Usage", "Aggregate Hours In Queue", "Aggregate Core Hours", "Number of Jobs"]
+    df = pd.DataFrame(columns = columnNames)
+    i = 0
+    for userAttr in queueTimes:
+        df.loc[i] = [userAttr.username, ",".join(usernameToProject[userAttr.username]), userAttr.starttime, userAttr.endtime, userAttr.idledays, userAttr.queuetime/HOUR, userAttr.corehours, userAttr.njobs]
+        i+= 1
+    
+    df = df.loc[df["Aggregate Core Hours"] > 0]
+
+    def replaceCN(user):
+        return user.replace("/OU=LocalUser/CN=", "")
+    df["Username"] = df["Username"].apply(replaceCN)
+
+    with open("output.csv", "w") as output_csv:
+        output_csv.write(df.to_csv(index=False))
+    print(df.to_csv(index=False))
+
+
+
+if __name__ == "__main__":
+    main()
```

Comment:
No newline at eof.

(It makes me a little sad that text editors even allow this.)